scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -471,7 +471,18 @@ trait TypedPipe[+T] extends Serializable {
    * Reasonably common shortcut for cases of total associative/commutative reduction
    * returns a ValuePipe with only one element if there is any input, otherwise EmptyValue.
    */
-  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = ComputedValue(groupAll.sum[U].values)
+  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = {
+    // every 1000 items, compact.
+    lazy implicit val batchedSG = Batched.compactingSemigroup[U](1000)
+    ComputedValue(map { t => ((), Batched[U](t)) }
+      .sumByLocalKeys
+      // remove the Batched before going to the reducers
+      .map { case (_, batched) => batched.sum }
```

Comment:
Keep in mind guys, in those instances we already know we only have 1 key.
Next recall that tuple semigroups/monoids already assume they can buffer
1000 items in memory.

This seems very safe to me given the default map-side cache size in
scalding 100K items (keys and values).

Do you think it poses more risk than the above two areas we buffer? (Also
recall sum on TypedPipe is a seldom used API).

On Monday, June 27, 2016, Avi Bryant <notifications@github.com> wrote:

> In scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala
> <https://github.com/twitter/scalding/pull/1575#discussion_r68626836>:
>
> > @@ -471,7 +471,18 @@ trait TypedPipe[+T] extends Serializable {
> >     * Reasonably common shortcut for cases of total associative/commutative reduction
> >     * returns a ValuePipe with only one element if there is any input, otherwise EmptyValue.
> >     */
> > -  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = ComputedValue(groupAll.sum[U].values)
> > +  def sum[U >: T](implicit plus: Semigroup[U]): ValuePipe[U] = {
> > +    // every 1000 items, compact.
> > +    lazy implicit val batchedSG = Batched.compactingSemigroup[U](1000)
> > +    ComputedValue(map { t => ((), Batched[U](t)) }
> > +      .sumByLocalKeys
> > +      // remove the Batched before going to the reducers
> > +      .map { case (_, batched) => batched.sum }
>
> (Deleted my previous comments because I wasn't thinking straight)
>
> It would not be difficult, I think, to write a single-key version of
> AdaptiveCache which used the same trick to tune the batch-size for Batched
> to make sure that we never ran out of memory.
>
> It's also interesting to think about how this optimization could be built,
> more generally, into AdaptiveCache itself - that is, it could try to grow
> the Batched size as well as the number of keys in the cache. Once it
> starts hitting memory pressure it can decide whether to tradeoff reducing
> the batch size to grow the number of keys further.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/scalding/pull/1575/files/ddc861052d386a5cd3a4f14a368f1bf2cfe5902f..6fcddab5be5c2b183707a75a51703551890ade5f#r68626836>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAEJdiU1AUVHOulADsol404MaSzw5aCrks5qQBIrgaJpZM4I-dkQ>
> .
>


-- 
P. Oscar Boykin, Ph.D. | http://twitter.com/posco | http://pobox.com/~boykin